### PR TITLE
Issue #4425: fixed multidimensional array types

### PR DESCRIPTION
**Pull request by sergileon**
Fixed [issue](https://github.com/checkstyle/checkstyle/issues/4425)

### CODE REVIEW

**Comment by romani** [comment]
@sergileon, please generate diff report.

Please add to input extra cases of Boolean in generic. And extra test illegal type "MyClass" and in input generic "MyClass<Runnable>" to make it clear how we work with generics.

Please also extend xdoc documentation on how exactly we match type. I did not find it at:
http://checkstyle.sourceforge.net/config_coding.html#IllegalType

**Comment by sergileon** [comment]
@romani , [diff report](https://sergileon.github.io/4425/index.html). Are you sure generics is work? If yes, help me with them, please. I create class:
```
private MyClass<Boolean> b;
private class MyClass<T>{
    T a;
    T[] b;
    T[][] c;
}
```
and added attribute: `type Boolean` for `illegalClassNames` - but it don't work.

**Comment by romani** [comment]
@sergileon , please always follow https://github.com/checkstyle/checkstyle/blob/master/.github/ISSUE_TEMPLATE.md when you want to share some example of behavior. It takes your extra minute but helps me to reply you from phone as all details are shown in such format.

**Comment by romani** [comment]
@sergileon ,
your case:
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <module name="TreeWalker">
        <module name="IllegalTypeCheck">
            <property name="illegalClassNames" value="Boolean"/>
        </module>
    </module>
</module>

$ cat MyClass.java 
public class MyClass {
    private MyClass1<Boolean> b;
    private class MyClass1<T>{
        T a;
        T[] b;
        T[][] c;
    }
}

$ java -jar checkstyle-8.2-all.jar -c config.xml MyClass.java 
Starting audit...
Audit done.
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeSameFileName.java [hunk 1]
```diff
@@ -26,4 +26,7 @@ private void foo() {
     }
     java.util.List<Integer> list = new ArrayList<>(); //WARNING
     private ArrayList<String> values;
+    private Boolean d;
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeSameFileName.java]: please put "//WARNING" to lines where violation is expected (see examples above).
This file is used for certain configuration, so it should be clear by reading Input file where violations are.

**Comment by sergileon** [comment]
@romani , [diff report](https://sergileon.github.io/4425/index.html)

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeSameFileName.java (hunk 1 above)
Comment (sergileon) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeSameFileName.java]: fixed

File: src/xdocs/config_coding.xml [hunk 2]
```diff
@@ -2010,7 +2010,7 @@ class SomeClass
         <p>Since Checkstyle 3.2</p>
         <p>
           Checks that particular classes are never used as types in variable
-          declarations, return values or parameters.
+          declarations, one dimensional and multi-dimensional arrays, return values or parameters.
```
Comment (Vladlis) [inline src/xdocs/config_coding.xml]: `one dimensional` -> `one-dimensional`

Comment (sergileon) [reply src/xdocs/config_coding.xml]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtils.java [hunk 3]
```diff
@@ -76,18 +76,15 @@ private CheckUtils() {
...
-    public static FullIdent createFullType(DetailAST typeAST) {
-        final DetailAST arrayDeclaratorAST =
-            typeAST.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
-        final FullIdent fullType;
+    public static FullIdent createFullType(final DetailAST typeAST) {
+        DetailAST ast = typeAST;
 
-        if (arrayDeclaratorAST == null) {
-            fullType = createFullTypeNoArrays(typeAST);
+        // ignore (multi-dimensional) array part of type
```
Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtils.java]: When I read this comment I start thinking that only multi-dimensional arrays are ignored. Please make it just `ignore (multi-dimensional) array part of type`.

Comment (sergileon) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CheckUtils.java]: fixed

**Comment by rnveach** [comment]
@sergileon You provided regression for `IllegalTypeCheck` but `CheckUtils.createFullType` is a common method among many checks.
https://github.com/checkstyle/checkstyle/blob/66cc9d43c9620678078d2716308ee61916227fcb/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java#L385
Please also provide regression on `AbstractClassCouplingCheck` and `ClassFanOutComplexityCheck`.

**Comment by romani** [comment]
@sergileon , 
diff report is good, code is fine.

please provide regression for affected Checks (see comment of rnveach). If no anomalies is found - I am ok merge.

**Comment by sergileon** [comment]
update [diff report](https://sergileon.github.io/4425/index.html). The branch master contain bug.
```
$ javac Test.java

$ cat Test.java

public class Test{
    protected Object[][] foo() {
        return new Object[][] {};
    }
}

$ cat config.xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name = "Checker">
    <module name="TreeWalker">
		    <module name="ClassFanOutComplexity">
			       <property name="max" value="0"/>
		    </module>
    </module>
</module>

$ java -jar checkstyle-8.3-SNAPSHOT-all.jar -c config.xml Test.java

Starting audit...
[ERROR] C:\Users\Ruslan_Iagudin\Desktop\New folder\Test.java:1:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 1 errors.
```
Expected 0 error. `Class Fan-Out Complexity is 1`, because multi-dimensional parsed with error. `createFullType `find `[` instead "object" for `new Object[][]`

**Comment by romani** [comment]
@sergileon ,

> The branch master contain bug.

please copy this report to separate issue.

**Comment by romani** [comment]
regression is weird.
But looks like it shows problems in ClassFanOutComplexity, that existed before this PR.
Example: https://sergileon.github.io/4425/apache-commons/index.html#A16
I do not understand why there was "3", as there is at least 5 imports that we do not exclude.

@sergileon , 
1) please provide explanation why we have regression "3" -> "2" in this case.
2) Is this case just another bug in Check ? if yes , please create issue and we will merge this PR.

**Comment by sergileon** [comment]
@romani, is "3" for `master`:
1. [ToStringBuilder](https://sergileon.github.io/4425/apache-commons/xref/project/check/contribution/contribution/checkstyle-tester/repositories/apache-commons/src/test/java/org/apache/commons/lang3/builder/DefaultToStringStyleTest.java.html#L41)
2. [Person](https://sergileon.github.io/4425/apache-commons/xref/project/check/contribution/contribution/checkstyle-tester/repositories/apache-commons/src/test/java/org/apache/commons/lang3/builder/DefaultToStringStyleTest.java.html#L86)
3. [ '\[' ](https://sergileon.github.io/4425/apache-commons/xref/project/check/contribution/contribution/checkstyle-tester/repositories/apache-commons/src/test/java/org/apache/commons/lang3/builder/DefaultToStringStyleTest.java.html#L123)
Is "2" for `patch_branch`:
1) [ToStringBuilder](https://sergileon.github.io/4425/apache-commons/xref/project/check/contribution/contribution/checkstyle-tester/repositories/apache-commons/src/test/java/org/apache/commons/lang3/builder/DefaultToStringStyleTest.java.html#L41)
2) [Person](https://sergileon.github.io/4425/apache-commons/xref/project/check/contribution/contribution/checkstyle-tester/repositories/apache-commons/src/test/java/org/apache/commons/lang3/builder/DefaultToStringStyleTest.java.html#L86)

**Comment by sergileon** [comment]
@romani, [new issue](https://github.com/checkstyle/checkstyle/issues/5134)

**Comment by romani** [comment]
@rnveach , please do final review

**Pull request merged**